### PR TITLE
T/115: Fix lists in converters

### LINKED ISSUE #115: CKEditor doesn't understand invalid nested lists
Evernote produces following HTML:
```
<ul>
    <li>item 1</li>
    <li>item 2</li>
    <ul>
        <li>sub-item 2.1</li>
        <li>sub-item 2.2</li>
        <ul>
            <li>sub-item 2.2.1</li>
        </ul>
    </ul>
</ul>
```

Obviously, this (ul -> ul instead of ul -> li -> ul) isn't valid, but browsers don't have any trouble rendering it as expected. But CKEditor does this:

image

So it just ignores the nested list altogether. I'm not sure what approach CKEditor takes regarding the invalid HTML.

### LINKED ISSUE #3007: CKEditor doesn't understand invalid nested lists
Evernote produces following HTML:
```
<ul>
    <li>item 1</li>
    <li>item 2</li>
    <ul>
        <li>sub-item 2.1</li>
        <li>sub-item 2.2</li>
        <ul>
            <li>sub-item 2.2.1</li>
        </ul>
    </ul>
</ul>
```

Obviously, this (ul -> ul instead of ul -> li -> ul) isn't valid, but browsers don't have any trouble rendering it as expected. But CKEditor does this:

image

So it just ignores the nested list altogether. I'm not sure what approach CKEditor takes regarding the invalid HTML.

### PR DESCRIPTION
**Pull request by jodator**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Fix: Convert invalid nested lists. Closes ckeditor/ckeditor5#3007.

---

### Additional information

* This makes changes in the PFO plugin redundant: https://github.com/ckeditor/ckeditor5-paste-from-office/pull/80
* I think that we could refactor how indents are calculated in general now we have indent + offset and this could be simplified to one (but maybe more complex) solution.
* This PR also fixes pasting from Gmail:

Peek 2019-08-14 11-39

Before:

Peek 2019-08-14 11-41

### CODE REVIEW

**Comment by jodator** [comment]
@Mgsy Could you check how this works? Should work with any invalid list created with other text processors like gmail list or google docs. So no content list item should be lost now.

**Review by Mgsy — APPROVED** [review]
LGTM 👍

**Comment by Reinmar** [comment]
I logged the final indent and the offset and they are always equal. So, this means that the tests you wrote cover only the case when `conversionStore.indent` equals `0`. It'd be good to cover also the case when it's `0+` because it will tell us whether the arithmetic is right.

image

**Comment by Reinmar** [comment]
I'd also think whether we can avoid to have `conversionStore.indent` and `conversionStore.indentOffsets` separately. Can't we somehow rely on just the indent? Maybe it should be modified whenever an incorrect list is detected?

**Review by Reinmar — CHANGES_REQUESTED** [review]
Looking for the way to simplify this code.

**Comment by jodator** [comment]
@Reinmar So I've added two more test cases as you suggested in 69befbc. After that, I've refactored the way the indent is calculated. Now we have a single function that calculates the indent. 

I think that there's a possibility to calculate an indent solely based on the list structure but OTOH I don't think that we need to put more work into this.

**Comment by jodator** [comment]
... and after writing the previous comment: https://github.com/ckeditor/ckeditor5-list/pull/141/commits/c5f3e8ca1073998c3a9e22814a53587aadab9df6.

I've made a POC of that and it worked - it's simplified enough. The idea behind this approach is that the indent depends on the number of parent `li` items (in the same ul > li tree). There's only one case where we want properly nest:
```
//		before:                           fixed list:
//		OL                                OL
//		|-> LI                            |-> LI         (number of parent LI: 0, indent: 0)
//		|-> OL                                |-> OL
//		    |-> LI                                |-> LI (number of parent LI: 0, indent: 1)
```

**Comment by Reinmar** [comment]
Hm... I like that the logic got simpler. But I'm a bit worried that we're removing a piece of code which perhaps did some magic that we'll now forget.

@oskarwrobel, AFAICS you're the author of the current conversion store based solution. Do you some danger in the solution which @jodator proposed or it's too long after you did your changes here?

Also, I'd be cool if you could re-test this a bit @Mgsy.

**Comment by oskarwrobel** [comment]
> @oskarwrobel, AFAICS you're the author of the current conversion store based solution. Do you some danger in the solution which @jodator proposed or it's too long after you did your changes here?

It was a long time ago :) Thankfully we have a @Mgsy and lot of unit tests :D 
I remember I wasn't satisfied with the conversion result of non-compliant HTML. Now it looks better.

File: src/converters.js [hunk 1]
```diff
@@ -984,11 +981,79 @@ function hoistNestedLists( nextIndent, modelRemoveStartPosition, viewRemoveStart
...
+//
+// @param {module:engine/view/element~Element} listItem
+// @param {Object} conversionStore
+// @returns {Number}
+function getIndent( listItem ) {
+	let indent = 0;
+
+	let parent = listItem.parent;
+
+	while ( parent && ( parent.is( 'li' ) || isList( parent ) ) ) {
```
Comment (Reinmar) [inline src/converters.js]: Regression:

```
<li>
    text
    <div>
        <ul>
             <li>foo</li>
        </ul>
     </div>
</li>
```

That's the danger of reading the indent from the tree without the context of what we had previously consumed.

If we'll iterate up to the top of the tree, then we may process stuff that got or will get filtered out. If we stop at ul/ol/li, then we ignore some scenarios.

So... I'd unforunately see safer if we relied on stuff that was already converted. In other words, on what we know really is in the model.

Comment (Reinmar) [reply src/converters.js]: I tested https://jsfiddle.net/tz8j5Ly6/ on master and it works fine. It does not work in this branch. Neither on HEAD, nor e.g. https://github.com/ckeditor/ckeditor5-list/pull/141/commits/1a701a08dfa419e875bd15313d36559cab91fbeb.

Comment (jodator) [reply src/converters.js]: So we definitively must add this case to the tests then. Looks like I've relied too much on current tests :( I'll look into this and similar problems:
a) no parent ul/ol
b) something wrapping inner ul/ol in the tree

I think that we _could_ calculate this from a tree but differently. But if not doable in a sensible way I'll try to make a solution based on the model (already converter LIs).

**Comment by jodator** [comment]
@Reinmar so checking how the nested lists are rendered when wrapped in other elements. It turned out that browsers are continuing nesting list (like ul > table > ul > table > ul ): https://jsfiddle.net/setor6mk/3/. So I've just removed the check for the parent being li/ul/ol and we go all the way up to the root :man_shrugging:.

I can add more tests though but I don't see anything more specific then that.

**Comment by jodator** [comment]
@Reinmar I've got tired by this ticket already. Please take a look at the added tests. Basically, the previous algorithm works well (the one that counts parent `<li>`) the only problem I had was a list inside a table inside a list.

This was converted OK visually (nested levels, etc) but failed in the model (the `listItem` in the `tableCell` got `indent=1`. It turned out that list post-fixer for indents didn't expect that case. I've added such case to the post-fixer and the list is now properly indendent

Still I don't see anything simple and realiable enough to fix that upcast conversion. I've tried something with the model but we cannot map previously converted items to check this on model level better. I think that post-fixer is good enough.

**Pull request merged**